homeassistant/components/luftdaten/__init__.py

```diff
@@ -0,0 +1,200 @@
+                    CONF_SHOW_ON_MAP: conf[CONF_SHOW_ON_MAP],
+                }
+            )
+        )
+
+    hass.data[DOMAIN][CONF_SCAN_INTERVAL] = conf[CONF_SCAN_INTERVAL]
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up Luftdaten as config entry."""
+    from luftdaten import Luftdaten
+    from luftdaten.exceptions import LuftdatenError
+
+    session = async_get_clientsession(hass)
+
+    try:
+        luftdaten = LuftDatenData(
+            Luftdaten(
+                config_entry.data[CONF_SENSOR_ID], hass.loop, session),
+            config_entry.data.get(CONF_SENSORS, {}).get(
+                    CONF_MONITORED_CONDITIONS, list(SENSORS)))
+        await luftdaten.async_update()
+        hass.data[DOMAIN][DATA_LUFTDATEN_CLIENT][config_entry.entry_id] = \
+            luftdaten
+    except LuftdatenError as err:
+        _LOGGER.error("An error occurred: %s", str(err))
+        hass.components.persistent_notification.create(
+            'Error: {0}<br />'
+            'You will need to restart Home Assistant after fixing.'
+            ''.format(err),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+        config_entry, 'sensor'))
+
+    async def refresh_sensors(event_time):
+        """Refresh Luftdaten data."""
+        await luftdaten.async_update()
+        async_dispatcher_send(hass, TOPIC_UPDATE)
+
+    hass.data[DOMAIN][DATA_LUFTDATEN_LISTENER][
+        config_entry.entry_id] = async_track_time_interval(
+            hass, refresh_sensors,
+            hass.data[DOMAIN].get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL))
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload an Luftdaten config entry."""
+    for component in ('sensor', ):
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    hass.data[DOMAIN][DATA_LUFTDATEN_CLIENT].pop(config_entry.entry_id)
+
+    remove_listener = hass.data[DOMAIN][DATA_LUFTDATEN_LISTENER].pop(
+        config_entry.entry_id)
+    remove_listener()
+
+    return True
+
+
+class LuftDatenData:
+    """Define a generic Luftdaten object."""
+
+    def __init__(self, client, sensor_conditions):
+        """Initialize the Luftdata object."""
+        self.client = client
+        self.data = {}
+        self.sensor_conditions = sensor_conditions
+
+    async def async_update(self):
+        """Update sensor/binary sensor data."""
+        from luftdaten.exceptions import LuftdatenError
+
+        try:
+            await self.client.get_data()
+
+            self.data[DATA_LUFTDATEN] = self.client.values
+            self.data[DATA_LUFTDATEN].update(self.client.meta)
+
+        except LuftdatenError:
+            _LOGGER.error("Unable to retrieve data from luftdaten.info")
+
+
+class LuftDatenEntity(Entity):
```

Comment:
Is this really necessary? It doesn't seem to really add value

Reply:
Removed